### PR TITLE
Abstract common Horovod operations from any specific framework

### PR DESCRIPTION
**Pull request by alsrgv**
In preparation for support of PyTorch, CNTK, or any other framework.

### CODE REVIEW

File: horovod/common/hashes.h [hunk 1]
```diff
@@ -33,6 +34,26 @@ template <typename T> struct hash<std::vector<T>> {
   }
 };
 
+template <typename U, typename V> struct hash<std::tuple<U, V>> {
+  typedef std::tuple<U, V> argument_type;
+  typedef std::size_t result_type;
+
+  result_type operator()(argument_type const& in) const {
+    size_t seed = 0;
+    seed ^= std::hash<U>()(std::get<0>(in)) + 0x9e3779b9 + (seed << 6) +
```
Comment (jeevandev) [inline horovod/common/hashes.h]: why this 0x9e3779b9 specific value? can we define it as a constant and not use the magic number?

File: horovod/common/__init__.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+
+def local_size():
+    """A function which returns the number of Horovod processes within the
+    node the current process is running on.
+
+    Returns:
+      An integer scalar containing the number of local Horovod processes.
+    """
+    local_size = MPI_LIB_CTYPES.horovod_local_size()
```
Comment (jeevandev) [inline horovod/common/__init__.py]: Where is this initialized in refactored code? I did not see

File: horovod/common/timeline.h [hunk 3]
```diff
@@ -20,26 +20,19 @@
 #include <fstream>
 #include <iostream>
 #include <unordered_map>
+#include <mutex>
+#include <atomic>
```
Comment (jeevandev) [inline horovod/common/timeline.h]: Why is mutex/atomic included?

File: horovod/common/__init__.py (hunk 2 above)
Comment (alsrgv) [reply horovod/common/__init__.py]: It's a function in `operations.cc`

File: horovod/common/operations.cc [hunk 4]
```diff
@@ -0,0 +1,1577 @@
...
+        CUDA_CHECK(
+            entries, "cudaStreamSynchronize",
+            cudaStreamSynchronize(horovod_global.streams[first_entry.device]))
+      }
+#endif
+      ACTIVITY_END_ALL(entries, timeline)
+    } else {
+      auto e = first_entry;
+      ACTIVITY_START_ALL(entries, timeline, "MPI_ALLREDUCE")
+      const void* sendbuf = e.tensor->data() == e.output->data()
```
Comment (alsrgv) [inline horovod/common/operations.cc]: Here we've added support for in-place allreduce.

File: horovod/common/hashes.h [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+} // namespace
+
+template <typename T> struct hash<std::vector<T>> {
+  typedef std::vector<T> argument_type;
+  typedef std::size_t result_type;
+
+  result_type operator()(argument_type const& in) const {
+    size_t size = in.size();
+    size_t seed = 0;
```
Comment (thepaulm) [inline horovod/common/hashes.h]: for consistency this should really be `result_type seed = 0;`

File: horovod/common/hashes.h [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    return seed;
+  }
+};
+
+template <typename U, typename V> struct hash<std::tuple<U, V>> {
+  typedef std::tuple<U, V> argument_type;
+  typedef std::size_t result_type;
+
+  result_type operator()(argument_type const& in) const {
+    size_t seed = 0;
```
Comment (thepaulm) [inline horovod/common/hashes.h]: this guy too

File: horovod/common/operations.cc [hunk 7]
```diff
@@ -0,0 +1,1577 @@
+// Copyright 2016 The TensorFlow Authors. All Rights Reserved.
```
Comment (thepaulm) [inline horovod/common/operations.cc]: How is this copyright tensorflow? did this file start from there? We should have Uber copyright?

File: horovod/common/operations.cc [hunk 8]
```diff
@@ -0,0 +1,1577 @@
...
+  } else if (message_type == MPIRequest::BROADCAST) {
+    response.set_response_type(MPIResponse::BROADCAST);
+  }
+  response.set_devices(devices);
+
+  // Clear all queued up requests for this name. They are now taken care of
+  // by the constructed MPI response.
+  message_table->erase(it);
+
+  return response;
```
Comment (thepaulm) [inline horovod/common/operations.cc]: A valid C++ pattern is to have like this:

```
return response;

error_return:
    return Failure;
```

and then instead of all of these early exit `if error` you can just do `goto error_return;`

File: horovod/common/operations.cc [hunk 9]
```diff
@@ -0,0 +1,1577 @@
...
+
+Status CheckInitialized() {
+  if (!horovod_global.initialization_done) {
+    return Status::PreconditionError(
+        "Horovod has not been initialized; use hvd.init().");
+  }
+  return Status::OK();
+}
+
+extern "C" void horovod_init() { InitializeHorovodOnce(); }
```
Comment (thepaulm) [inline horovod/common/operations.cc]: You can make one big `extern "C"` block here ...

File: horovod/common/operations.cc (hunk 7 above)
Comment (alsrgv) [reply horovod/common/operations.cc]: We do have Uber copyright there as well.
